addons/html_editor/static/src/editor/core/history_plugin.js

```diff
@@ -2,37 +2,585 @@
 
 import { registry } from "@web/core/registry";
 import { Plugin } from "../plugin";
+import { setSelection } from "../utils/selection";
+import { serializeSelection } from "../utils/serialize";
 
 export class HistoryPlugin extends Plugin {
     static name = "history";
-    static shared = ["enableObserver", "disableObserver"];
+    static dependencies = ["dom"];
+
+    // @todo @phoenix usefull for the collaboration plugin. See if we still want
+    // to handle the serialization through that property later.
+    isSerializable = false;
 
     setup() {
-        this.observer = new MutationObserver(() => this.handleDOMChange());
-        this.enableObserver();
+        this.registry.category("shortcuts").add("control+z", { command: "HISTORY_UNDO" });
+        this.registry.category("shortcuts").add("control+y", { command: "HISTORY_REDO" });
+        this.addDomListener(this.editable, "input", this.addStep);
+        this.addDomListener(this.editable, "keydown", this.stageSelection);
+        this.addDomListener(this.editable, "beforeinput", this.stageSelection);
+        this.observer = new MutationObserver(this.handleNewRecords.bind(this));
         this._cleanups.push(() => this.observer.disconnect());
     }
+    start() {
+        this.enableObserver();
+        this.reset();
+        this.stageFirstStep();
+        this.renderingClasses = new Set(
+            this.registry.category("history_rendering_classes").getAll().flat()
+        );
+    }
+    handleCommand(command, payload) {
+        switch (command) {
+            case "HISTORY_UNDO":
+                this.undo();
+                break;
+            case "HISTORY_REDO":
+                this.redo();
+                break;
+            case "ADD_STEP":
+                this.addStep();
+                break;
+            case "HISTORY_STAGE_SELECTION":
+                this.stageSelection();
+                break;
+        }
+    }
+
+    reset() {
+        this.steps = [];
+        this.currentStep = {
+            selection: {
+                anchorNodeOid: undefined,
+                anchorOffset: undefined,
+                focusNodeOid: undefined,
+                focusOffset: undefined,
+            },
+            mutations: [],
+            id: undefined,
+            clientId: undefined,
+        };
+        this.stepsStates = new Map();
+        this.nodeToIdMap = new WeakMap();
+        this.idToNodeMap = new Map();
+        this.latestSelection = {
+            anchorNode: this.editable,
+            anchorOffset: 0,
+            focusNode: this.editable,
+            focusOffset: 0,
+        };
+        // @todo @phoenix add in collaboration plugin? previously _historyIds
+        this.branchStepIds = [];
+        this.setNodeId(this.editable);
+    }
+    stageFirstStep() {
```

Comment:
Is this meant to be called only once? Why is it in its own function?

Reply:
Because with Géry we relaxed the rule about having function being called only once.
It's self documenting and make the start more readable now and in the future.